File: src/iOS/Upload/UploadViewController.m
//
//  UploadViewController.m
//  Go Map!!
//
//  Created by Bryce Cogswell on 12/19/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

#import <MessageUI/MessageUI.h>
#import <QuartzCore/QuartzCore.h>

#if TARGET_OS_IPHONE
#import "DDXML.h"
#endif

#import "AerialList.h"
#import "AppDelegate.h"
#import "MapView.h"
#import "EditorMapLayer.h"
#import "MercatorTileLayer.h"
#import "OsmMapData.h"
#import "UploadViewController.h"


@implementation UploadViewController


- (void)viewDidLoad
{
    [super viewDidLoad];

	UIColor * color = [UIColor.grayColor colorWithAlphaComponent:0.5];
	_commentContainerView.layer.borderColor = color.CGColor;
	_commentContainerView.layer.borderWidth = 2.0;
	_commentContainerView.layer.cornerRadius = 10.0;

	_sourceTextField.layer.borderColor = color.CGColor;
	_sourceTextField.layer.borderWidth = 2.0;
	_sourceTextField.layer.cornerRadius = 10.0;

	_xmlTextView.layer.borderColor = color.CGColor;
	_xmlTextView.layer.borderWidth = 2.0;
	_xmlTextView.layer.cornerRadius = 10.0;
}

- (void)viewWillAppear:(BOOL)animated
{
	[super viewWillAppear:animated];

	MapView * mapView = [(AppDelegate *)[[UIApplication sharedApplication] delegate] mapView];
	_mapData = mapView.editorLayer.mapData;

	NSString * comment = [[NSUserDefaults standardUserDefaults] objectForKey:@"uploadComment"];
	_commentTextView.text = comment;

	NSString * source = [[NSUserDefaults standardUserDefaults] objectForKey:@"uploadSource"];
	_sourceTextField.text = source;

	NSAttributedString * text = [_mapData changesetAsAttributedString];
	if ( text == nil ) {
		_commitButton.enabled = NO;
		UIFont * font = [UIFont preferredFontForTextStyle:UIFontTextStyleBody];
		_xmlTextView.attributedText = [[NSAttributedString alloc] initWithString:NSLocalizedString(@"Nothing to upload, no changes have been made.",nil) attributes:@{ NSFontAttributeName : font }];
	} else {
		_commitButton.enabled = YES;
		_xmlTextView.attributedText = text;
	}

	_sendMailButton.enabled = (text != nil);
	_editXmlButton.enabled = (text != nil);

	_clearCommentButton.hidden = YES;
}

- (void)viewDidDisappear:(BOOL)animated
{
	[super viewDidDisappear:animated];
	[[NSUserDefaults standardUserDefaults] setObject:_commentTextView.text forKey:@"uploadComment"];
	[[NSUserDefaults standardUserDefaults] setObject:_sourceTextField.text forKey:@"uploadSource"];
}

- (BOOL)textView:(UITextView *)textView shouldChangeTextInRange:(NSRange)range replacementText:(NSString *)text
{
    if ( [text isEqualToString:@"\n"] ) {
		[textView resignFirstResponder];
		return NO;
	}
	return YES;
}


- (IBAction)clearCommentText:(id)sender
{
	_commentTextView.text = @"";
	_clearCommentButton.hidden = YES;
}


- (IBAction)commit:(id)sender
{
	AppDelegate * appDelegate = (id)[[UIApplication sharedApplication] delegate];
	if ( appDelegate.userName.length == 0 || appDelegate.userPassword.length == 0 ) {
		[self performSegueWithIdentifier:@"loginSegue" sender:self];
		return;
	}

	if ( ![[NSUserDefaults standardUserDefaults] boolForKey:@"userDidPreviousUpload"] ) {
		UIAlertController * alert = [UIAlertController alertControllerWithTitle:@"Attention"
																		message:@"You are about to make changes to the live OpenStreetMap database. Your changes will be visible to everyone in the world.\n\nTo continue press Commit once again, otherwise press Cancel."
																 preferredStyle:UIAlertControllerStyleAlert];
		[alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Cancel",nil) style:UIAlertActionStyleCancel handler:nil]];
		[alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"Commit",nil) style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
			[[NSUserDefaults standardUserDefaults] setBool:YES forKey:@"userDidPreviousUpload"];
			[self commit:nil];
		}]];
		[self presentViewController:alert animated:YES completion:nil];
		return;
	}

	_mapData.credentialsUserName = appDelegate.userName;
	_mapData.credentialsPassword = appDelegate.userPassword;

	[_progressView startAnimating];
	[_commitButton setEnabled:NO];
	[_cancelButton setEnabled:NO];
	[_sendMailButton setEnabled:NO];
	[_editXmlButton setEnabled:NO];

	[_commentTextView resignFirstResponder];
	[_xmlTextView resignFirstResponder];

	NSString * comment = _commentTextView.text;
	comment = [comment stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];

	NSString * source = _sourceTextField.text;
	source = [source stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];

	void (^completion)(NSString *) = ^void(NSString * error) {
		[_progressView stopAnimating];
		[_commitButton setEnabled:YES];
		[_cancelButton setEnabled:YES];
		if ( error ) {
			UIAlertController * alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Unable to upload changes",nil)
																			message:error
																	 preferredStyle:UIAlertControllerStyleAlert];
			[alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK",nil) style:UIAlertActionStyleCancel handler:nil]];
			[self presentViewController:alert animated:YES completion:nil];

			if ( !_xmlTextView.editable ) {
				[_sendMailButton setEnabled:YES];
				[_editXmlButton setEnabled:YES];
			}
		} else {

			[self dismissViewControllerAnimated:YES completion:nil];

			// flash success message
			dispatch_time_t popTime = dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.3 * NSEC_PER_SEC));
			dispatch_after(popTime, dispatch_get_main_queue(), ^(void){
				[appDelegate.mapView flashMessage:NSLocalizedString(@"Upload complete!",nil) duration:1.5];

				// record number of uploads
				NSString * uploadKey = [NSString stringWithFormat:@"uploadCount-%@", appDelegate.appVersion];
				NSInteger editCount = [[NSUserDefaults standardUserDefaults] integerForKey:uploadKey];
				++editCount;
				[[NSUserDefaults standardUserDefaults] setInteger:editCount forKey:uploadKey];
				[appDelegate.mapView askToRate:editCount];
			});
		}
	};

	NSString * imagery = nil;
	if ( appDelegate.mapView.viewState == MAPVIEW_EDITORAERIAL || appDelegate.mapView.viewState == MAPVIEW_AERIAL )
		imagery = appDelegate.mapView.aerialLayer.aerialService.name;

	if ( _xmlTextView.editable ) {
		
		// upload user-edited text
		NSString * xmlText = _xmlTextView.text;
		NSError * error = nil;
		NSXMLDocument * xmlDoc = [[NSXMLDocument alloc] initWithXMLString:xmlText options:0 error:&error];
		if ( error ) {
			completion( NSLocalizedString( @"The XML is improperly formed", nil ) );
			return;
		}
		[_mapData uploadChangesetXml:xmlDoc comment:comment source:source imagery:imagery completion:completion];

	} else {
		// normal upload
		[_mapData uploadChangesetWithComment:comment source:source imagery:imagery completion:completion];
	}
}

-(IBAction)editXml:(id)sender
{
	AppDelegate * appDelegate = AppDelegate.getAppDelegate;

	MFMailComposeViewController * mail = [[MFMailComposeViewController alloc] init];
	mail.mailComposeDelegate = self;
	[mail setSubject:[NSString stringWithFormat:@"%@ changeset", appDelegate.appName]];
	NSString * xml = [_mapData changesetAsXml];
	xml = [xml stringByAppendingString:@"\n\n\n\n\n\n\n\n\n\n\n\n"];
	_xmlTextView.attributedText = nil;
	_xmlTextView.text = xml;
	_xmlTextView.editable = YES;
	_sendMailButton.enabled = NO;
	_editXmlButton.enabled = NO;

	UIAlertController * alert = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Edit XML",nil)
																	message:NSLocalizedString(@"Modifying the raw XML data allows you to correct errors that prevent uploading.\n\nIt is an advanced operation that should only be undertaken if you have a thorough understanding of the OSM changeset format.",nil)
															 preferredStyle:UIAlertControllerStyleAlert];
	[alert addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK",nil) style:UIAlertActionStyleCancel handler:nil]];
	[self presentViewController:alert animated:YES completion:nil];
}

-(IBAction)sendMail:(id)sender
{
	if ( [MFMailComposeViewController canSendMail] ) {
		AppDelegate * appDelegate = (id)[[UIApplication sharedApplication] delegate];

		MFMailComposeViewController * mail = [[MFMailComposeViewController alloc] init];
		mail.mailComposeDelegate = self;
		[mail setSubject:[NSString stringWithFormat:@"%@ changeset", appDelegate.appName]];
		NSString * xml = [_mapData changesetAsXml];
		[mail addAttachmentData:[xml dataUsingEncoding:NSUTF8StringEncoding] mimeType:@"application/xml" fileName:@"osmChange.osc"];
		[self presentViewController:mail animated:YES completion:nil];
	} else {
		UIAlertController * error = [UIAlertController alertControllerWithTitle:NSLocalizedString(@"Cannot compose message",nil)
																		message:NSLocalizedString(@"Mail delivery is not available on this device",nil)
																 preferredStyle:UIAlertControllerStyleAlert];
		[error addAction:[UIAlertAction actionWithTitle:NSLocalizedString(@"OK",nil) style:UIAlertActionStyleCancel handler:nil]];
		[self presentViewController:error animated:YES completion:nil];
	}
}

- (void)mailComposeController:(MFMailComposeViewController*)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError*)error
{
	[self dismissViewControllerAnimated:YES completion:nil];
}

-(void)textViewDidChange:(UITextView *)textView
{
	if ( textView == _commentTextView ) {
		_clearCommentButton.hidden  = _commentTextView.text.length == 0;
	}
}
-(void)textViewDidBeginEditing:(UITextView *)textView
{
	if ( textView == _commentTextView ) {
		_clearCommentButton.hidden  = _commentTextView.text.length == 0;
	}
}
-(void)textViewDidEndEditing:(UITextView *)textView
{
	if ( textView == _commentTextView ) {
		_clearCommentButton.hidden  = YES;
	}
}

// this is for navigating from the changeset back to the location of the modified object
- (BOOL)textView:(UITextView *)textView shouldInteractWithURL:(NSURL *)url inRange:(NSRange)characterRange
{
	AppDelegate	*	appDelegate = AppDelegate.getAppDelegate;
	NSString 	*	name = url.absoluteString;
	if ( name.length == 0 )
		return NO;
	OsmIdentifier 	ident = [[name substringFromIndex:1] longLongValue];
	switch ( [name characterAtIndex:0] ) {
		case 'n':
			ident = [OsmBaseObject extendedIdentifierForType:OSM_TYPE_NODE identifier:ident];
			break;
		case 'w':
			ident = [OsmBaseObject extendedIdentifierForType:OSM_TYPE_WAY identifier:ident];
			break;
		case 'r':
			ident = [OsmBaseObject extendedIdentifierForType:OSM_TYPE_RELATION identifier:ident];
			break;
		default:
			return NO;
	}
	OsmBaseObject *	object = [appDelegate.mapView.editorLayer.mapData objectWithExtendedIdentifier:@(ident)];
	if ( object == nil )
		return NO;
	
	appDelegate.mapView.editorLayer.selectedRelation = object.isRelation;
	appDelegate.mapView.editorLayer.selectedWay		 = object.isWay;
	appDelegate.mapView.editorLayer.selectedNode	 = object.isNode;
	[appDelegate.mapView placePushpinForSelection];
	
	[self cancel:nil];
	return NO;
}

- (IBAction)cancel:(id)sender
{
	[self dismissViewControllerAnimated:YES completion:nil];
}

@end

File: src/iOS/Upload/UploadViewController.swift
//
//  UploadViewController.swift
//  Go Map!!
//
//  Created by Bryce Cogswell on 12/19/12.
//  Copyright (c) 2012 Bryce Cogswell. All rights reserved.
//

import MessageUI
import QuartzCore
import UIKit

func sanitizedURL(_ urlString: String) -> String {
	let sensitiveKeys = ["token", "auth", "api_key", "access_token", "connectid", "signature"]

	guard var components = URLComponents(string: urlString) else { return urlString }
	components.queryItems = components.queryItems?.map { item in
		if sensitiveKeys.contains(item.name.lowercased()) {
			return URLQueryItem(name: item.name, value: "{apikey}")
		} else {
			return item
		}
	}
	return components.url?.absoluteString ?? urlString
}

class UploadViewController: UIViewController, UITextViewDelegate {
	var mapData: OsmMapData!
	@IBOutlet var commentContainerView: UIView!
	@IBOutlet var xmlTextView: UITextView!
	@IBOutlet var commentTextView: UITextView!
	@IBOutlet var sourceTextField: UITextField!
	@IBOutlet var commitButton: UIBarButtonItem!
	@IBOutlet var cancelButton: UIBarButtonItem!
	@IBOutlet var progressView: UIActivityIndicatorView!
	@IBOutlet var exportOscButton: UIButton!
	@IBOutlet var editXmlButton: UIButton!
	@IBOutlet var clearCommentButton: UIButton!
	@IBOutlet var commentHistoryButton: UIButton!
	@IBOutlet var sourceHistoryButton: UIButton!
	@IBOutlet var changesetCommentPlaceholder: UILabel!

	var recentCommentList = MostRecentlyUsed<String>(maxCount: 5,
	                                                 userPrefsKey: UserPrefs.shared.recentCommitComments)
	var recentSourceList = MostRecentlyUsed<String>(maxCount: 5,
	                                                userPrefsKey: UserPrefs.shared.recentSourceComments)

	override func viewDidLoad() {
		super.viewDidLoad()

		let color = UIColor.gray.withAlphaComponent(0.5)
		commentContainerView.layer.borderColor = color.cgColor
		commentContainerView.layer.borderWidth = 2.0
		commentContainerView.layer.cornerRadius = 10.0

		sourceTextField.layer.borderColor = color.cgColor
		sourceTextField.layer.borderWidth = 2.0
		sourceTextField.layer.cornerRadius = 10.0

		xmlTextView.layer.borderColor = color.cgColor
		xmlTextView.layer.borderWidth = 2.0
		xmlTextView.layer.cornerRadius = 10.0

		// create button for source history
		sourceHistoryButton = UIButton(type: .custom)
		sourceHistoryButton.frame = CGRect(x: 0, y: 0, width: 22, height: 22)
		sourceHistoryButton.setTitle("🔽", for: .normal)
		sourceHistoryButton.addTarget(self, action: #selector(showSourceHistory), for: .touchUpInside)
		sourceTextField.rightView = sourceHistoryButton
		sourceTextField.rightViewMode = .always

		if #available(iOS 13.0, *) {
			progressView.style = .large
		} else {
			progressView.style = .whiteLarge
		}
	}

	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)

		let mapView = AppDelegate.shared.mapView
		mapData = mapView?.editorLayer.mapData

		commentTextView.text = UserPrefs.shared.uploadComment.value
		sourceTextField.text = UserPrefs.shared.uploadSource.value
		sourceTextField.placeholder = "survey, Bing, knowledge" // overrules translations: see #557
		changesetCommentPlaceholder.isHidden = commentTextView.text.count > 0

		let text = mapData?.changesetAsAttributedString()
		if text == nil {
			commitButton.isEnabled = false
			let font = UIFont.preferredFont(forTextStyle: .body)
			xmlTextView.attributedText = NSAttributedString(
				string: NSLocalizedString("Nothing to upload, no changes have been made.", comment: ""),
				attributes: [
					NSAttributedString.Key.font: font
				])
		} else {
			commitButton.isEnabled = true
			xmlTextView.attributedText = text
		}

		exportOscButton.isEnabled = text != nil
		editXmlButton.isEnabled = text != nil

		clearCommentButton.isHidden = true

		commentHistoryButton.isHidden = recentCommentList.count == 0
		sourceTextField.rightViewMode = recentSourceList.count > 0 ? .always : .never
	}

	override func viewWillDisappear(_ animated: Bool) {
		super.viewWillDisappear(animated)
		UserPrefs.shared.uploadComment.value = commentTextView.text
		UserPrefs.shared.uploadSource.value = sourceTextField.text
	}

	func textView(_ textView: UITextView, shouldChangeTextIn range: NSRange, replacementText text: String) -> Bool {
		if text == "\n" {
			textView.resignFirstResponder()
			return false
		}
		return true
	}

	@IBAction func clearCommentText(_ sender: Any) {
		commentTextView.text = ""
		clearCommentButton.isHidden = true
	}

	private func showHistorySheet(_ list: [String], button: UIButton, textView: UIView) {
		let actionSheet = UIAlertController(
			title: nil,
			message: nil,
			preferredStyle: .actionSheet)
		for message in list {
			actionSheet.addAction(UIAlertAction(title: message, style: .default, handler: { _ in
				if let view = textView as? UITextView {
					view.text = message
					view.resignFirstResponder()
				} else if let view = textView as? UITextField {
					view.text = message
					view.resignFirstResponder()
				}
				self.changesetCommentPlaceholder.isHidden = self.commentTextView.text.count > 0
			}))
		}
		actionSheet.addAction(UIAlertAction(title: NSLocalizedString("Cancel", comment: ""),
		                                    style: .cancel,
		                                    handler: nil))
		actionSheet.popoverPresentationController?.sourceView = button
		actionSheet.popoverPresentationController?.sourceRect = button.bounds
		present(actionSheet, animated: true)
	}

	@IBAction func showCommitMessageHistory(_ sender: Any) {
		showHistorySheet(recentCommentList.items, button: commentHistoryButton, textView: commentTextView)
	}

	@IBAction func showSourceHistory(_ sender: Any) {
		showHistorySheet(recentSourceList.items, button: sourceHistoryButton, textView: sourceTextField)
	}

	@IBAction func commit(_ sender: Any?) {
		let appDelegate = AppDelegate.shared
		guard let oAuth = OSM_SERVER.oAuth2,
		      oAuth.isAuthorized()
		else {
			performSegue(withIdentifier: "loginSegue", sender: self)
			return
		}

		guard
			let didPrev = UserPrefs.shared.userDidPreviousUpload.value,
			didPrev
		else {
			let alert = UIAlertController(
				title: NSLocalizedString("Attention", comment: ""),
				message: NSLocalizedString(
					"You are about to make changes to the live OpenStreetMap database. Your changes will be visible to everyone in the world.\n\nTo continue press Commit once again, otherwise press Cancel.",
					comment: ""),
				preferredStyle: .alert)
			alert.addAction(UIAlertAction(title: NSLocalizedString("Cancel", comment: ""),
			                              style: .cancel,
			                              handler: nil))
			alert.addAction(UIAlertAction(title: NSLocalizedString("Commit", comment: ""),
			                              style: .default,
			                              handler: { [self] _ in
			                              	UserPrefs.shared.userDidPreviousUpload.value = true
			                              	commit(nil)
			                              }))
			present(alert, animated: true)
			return
		}

		progressView.startAnimating()
		commitButton.isEnabled = false
		cancelButton.isEnabled = false
		exportOscButton.isEnabled = false
		editXmlButton.isEnabled = false

		commentTextView.resignFirstResponder()
		xmlTextView.resignFirstResponder()

		var comment = commentTextView.text ?? ""
		comment = comment.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
		if comment != "" {
			recentCommentList.updateWith(comment)
		}

		var source = sourceTextField.text ?? ""
		source = source.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
		if source != "" {
			recentSourceList.updateWith(source)
		}

		let locale = PresetLanguages.preferredLanguageCode()

		let completion: ((Error?) -> Void) = { [self] error in
			progressView.stopAnimating()
			commitButton.isEnabled = true
			cancelButton.isEnabled = true
			if let error = error as? UrlSessionError,
			   case let .badStatusCode(code, _) = error,
			   code == 401
			{
				// authentication error, so redirect to login page
				OSM_SERVER.oAuth2?.removeAuthorization()
				performSegue(withIdentifier: "loginSegue", sender: self)
				return
			} else if let error = error {
				let alert = UIAlertController(
					title: NSLocalizedString("Unable to upload changes", comment: ""),
					message: error.localizedDescription,
					preferredStyle: .alert)
				alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""),
				                              style: .cancel,
				                              handler: nil))
				present(alert, animated: true)

				if !xmlTextView.isEditable {
					exportOscButton.isEnabled = true
					editXmlButton.isEnabled = true
				}
			} else {
				dismiss(animated: true)

				// flash success message
				MainActor.runAfter(nanoseconds: 300_000000) {
					appDelegate.mapView.editorLayer.setNeedsLayout()
					appDelegate.mapView.flashMessage(title: nil,
					                                 message: NSLocalizedString("Upload complete!", comment: ""),
					                                 duration: 1.5)

					// record number of uploads
					var editCount = UserPrefs.shared.uploadCountPerVersion.value ?? 0
					editCount += 1
					UserPrefs.shared.uploadCountPerVersion.value = editCount
					appDelegate.mapView.ask(toRate: editCount)
				}
			}
		}

		var imagery = ""
		if appDelegate.mapView.viewState == MapViewState.EDITORAERIAL ||
			appDelegate.mapView.viewState == MapViewState.AERIAL
		{
			let server = appDelegate.mapView.aerialLayer.tileServer
			if server.identifier.hasPrefix("http:") || server.identifier.hasPrefix("https:") {
				// custom user imagery
				imagery = sanitizedURL(server.identifier)
			} else {
				imagery = server.name
			}
		}

		if xmlTextView.isEditable {
			// upload user-edited text
			let xmlText = xmlTextView.text ?? ""
			let xmlDoc: DDXMLDocument
			do {
				xmlDoc = try DDXMLDocument(xmlString: xmlText, options: 0)
			} catch {
				completion(error)
				return
			}
			mapData?.openChangesetAndUpload(
				xml: xmlDoc,
				comment: comment,
				source: source,
				imagery: imagery,
				locale: locale,
				completion: completion)
		} else {
			// normal upload
			mapData?.uploadChangeset(withComment: comment,
			                         source: source,
			                         imagery: imagery,
			                         locale: locale,
			                         completion: completion)
		}
	}

	@IBAction func editXml(_ sender: Any) {
		var xml = mapData?.changesetAsXml() ?? ""
		xml = xml + "\n\n\n\n\n\n\n\n\n\n\n\n"
		xmlTextView.attributedText = nil
		xmlTextView.text = xml
		xmlTextView.isEditable = true
		exportOscButton.isEnabled = false
		editXmlButton.isEnabled = false

		let alert = UIAlertController(
			title: NSLocalizedString("Edit XML", comment: ""),
			message: NSLocalizedString(
				"Modifying the raw XML data allows you to correct errors that prevent uploading.\n\nIt is an advanced operation that should only be undertaken if you have a thorough understanding of the OSM changeset format.",
				comment: ""),
			preferredStyle: .alert)
		alert.addAction(UIAlertAction(title: NSLocalizedString("OK", comment: ""), style: .cancel, handler: nil))
		present(alert, animated: true)
	}

	@IBAction func exportOscFile(_ sender: Any) {
		if let xml = mapData?.changesetAsXml(),
		   let text = xml.data(using: .utf8),
		   let path = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first?
		   .appendingPathComponent("osmChange.osc"),
		   (try? text.write(to: path, options: .atomicWrite)) != nil
		{
			let objectsToShare = [path] as [Any]
			let activityVC = UIActivityViewController(activityItems: objectsToShare, applicationActivities: nil)

			// Excluded activities
			activityVC.excludedActivityTypes = [UIActivity.ActivityType.addToReadingList]

			activityVC.popoverPresentationController?.sourceView = sender as? UIView
			present(activityVC, animated: true, completion: nil)
		}
	}

	func textViewDidChange(_ textView: UITextView) {
		if textView == commentTextView {
			clearCommentButton.isHidden = commentTextView.text.count == 0
			changesetCommentPlaceholder.isHidden = commentTextView.text.count > 0
		}
	}

	func textViewDidBeginEditing(_ textView: UITextView) {
		if textView == commentTextView {
			clearCommentButton.isHidden = commentTextView.text.count == 0
			changesetCommentPlaceholder.isHidden = true
		}
	}

	func textViewDidEndEditing(_ textView: UITextView) {
		if textView == commentTextView {
			clearCommentButton.isHidden = true
			changesetCommentPlaceholder.isHidden = commentTextView.text.count > 0
		}
	}

	// this is for navigating from the changeset back to the location of the modified object
	func textView(
		_ textView: UITextView,
		shouldInteractWith url: URL,
		in characterRange: NSRange,
		interaction: UITextItemInteraction) -> Bool
	{
		let appDelegate = AppDelegate.shared
		let name = url.absoluteString
		if name.count == 0 {
			return false
		}
		let ident = Int64(name.dropFirst()) ?? 0
		let extendedId: OsmExtendedIdentifier
		switch name.prefix(1) {
		case "n":
			extendedId = OsmExtendedIdentifier(.NODE, ident)
		case "w":
			extendedId = OsmExtendedIdentifier(.WAY, ident)
		case "r":
			extendedId = OsmExtendedIdentifier(.RELATION, ident)
		default:
			return false
		}
		guard let object = appDelegate.mapView.editorLayer.mapData.object(withExtendedIdentifier: extendedId)
		else { return false }
		appDelegate.mapView.editorLayer.selectedRelation = object.isRelation()
		appDelegate.mapView.editorLayer.selectedWay = object.isWay()
		appDelegate.mapView.editorLayer.selectedNode = object.isNode()
		appDelegate.mapView.placePushpinForSelection()
		cancel(nil)
		return false
	}

	@IBAction func cancel(_ sender: Any?) {
		dismiss(animated: true)
	}
}
